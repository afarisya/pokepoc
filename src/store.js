import { createStore, applyMiddleware, compose } from 'redux';
import { routerMiddleware } from 'react-router-redux';
import thunk from 'redux-thunk';
// import WebsocketMiddleware from './middleware/WebsocketMiddleware';
import { createBrowserHistory, createHashHistory } from 'history';
import rootReducer from './reducers';

export const history = createBrowserHistory()

// Use this if you want to use routing with hash
// export const history = createHashHistory();

const initialState = {}
const enhancers = []
const middleware = [
  thunk,
  // logger,
  routerMiddleware(history)//,
//   WebsocketMiddleware
]

if (process.env.NODE_ENV === 'development') {
  const devToolsExtension = window.devToolsExtension

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension())
  }
}

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers
)

const store = createStore(
  rootReducer,
  initialState,
  composedEnhancers
)

export default store