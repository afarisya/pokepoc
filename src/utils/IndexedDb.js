// import store from '../store'
// import { setIndexedDB } from "../reducers/GlobalReducers";
// import { setUserPicture } from "../reducers/ProfileReducers";
// import { sndAudioStream, reqInitSendBlobChange, sndEosStream, sndEndStream } from '../reducers/NewMeetingReducers';

// import defaultPicture from '../img/admin.png';


export var db;

// IndexedDB
var indexedDB = window.indexedDB || window.webkitIndexedDB || window.mozIndexedDB || window.OIndexedDB || window.msIndexedDB;
// var IDBTransaction = window.IDBTransaction || window.webkitIDBTransaction || window.OIDBTransaction || window.msIDBTransaction;
var dbVersion = 1.0;

// Create/open database
var request = indexedDB.open("MyPokemonList", dbVersion);

var lastIndex = null;

request.onsuccess = function (event) {
    console.log("Success creating/accessing IndexedDB database");

    db = request.result;

    db.onerror = function (event) {
        console.log("Error creating/accessing IndexedDB database");
    };
    
}

request.onerror = function (event) {
    console.log("Error creating/accessing IndexedDB database");
};

// For future use. Currently only in latest Firefox versions
request.onupgradeneeded = function (event) {
    // console.log("Upgrading IndexedDb");
    var dataBase = event.target.result;

    var store = dataBase.createObjectStore("pokemons", { 
                    keyPath: 'pokemonId'//,
                    // autoIncrement: true 
                });
    store.createIndex("pokemonId", "pokemonId", { unique: true });
    store.createIndex("pokemonName", "pokemonName", { unique: false });
    store.createIndex("pokemonNickname", "pokemonNickname", { unique: false });

};

export var createObjectStore = function (dataBase, objStoreName) {
    // Create an objectStore
    // console.log("Creating objectStore")
    dataBase.createObjectStore(objStoreName);
};

export var getMyPokemonList = function(offset, limit){
    return new Promise((resolve, reject) => {
        var objectStore = db.transaction("pokemons").objectStore("pokemons");
        var pokemons = [];

        var iterate = 0;

        var request = objectStore.openCursor();

        request.onsuccess = function(event) {
            var cursor = event.target.result;
            if (cursor) {
                if ( iterate >= offset && pokemons.length < limit ) {
                    pokemons.push(cursor.value);
                }

                console.log(pokemons)
                if ( pokemons.length === limit ) {
                    getMyPokemonCount()
                        .then((count) => {
                            resolve(
                                {
                                    count: iterate,
                                    pokemons: pokemons
                                }
                            );
                        })
                        .catch((err) => {
                            reject("error")
                        })
                }

                iterate+=1;
                cursor.continue();

            } else {
                getMyPokemonCount()
                    .then((count) => {
                        resolve(
                            {
                                count: iterate,
                                pokemons: pokemons
                            }
                        );
                    })
                    .catch((err) => {
                        reject("error")
                    })
            }
        }

        request.onerror = function(event) {
            reject("error")
        }
    })


    // IDBKeyRange.bound(0, N, false, true);
}

export var getMyPokemonCount = function(){
    return new Promise((resolve, reject) => {
        var transaction = db.transaction(['pokemons'], 'readonly');
        var objectStore = transaction.objectStore('pokemons');
            
        var countRequest = objectStore.count();

        countRequest.onsuccess = function() {
            console.log(countRequest.result)
            resolve(countRequest.result);
        }

        countRequest.onerror = function(event) {
            reject("error");
        };
    })
}

export var getPokemonById = function(id){
    return new Promise((resolve, reject) => {
        var transaction = db.transaction(["pokemons"]);
        var objectStore = transaction.objectStore("pokemons");
        var request = objectStore.get(id);
        
        request.onsuccess = function(event) {
            // Do something with the request.result!
            resolve(request.result);
        };

        request.onerror = function(event) {
            reject("error");
        };
    })
}

export var addPokemon = function(obj){    
    return new Promise((resolve, reject) => {
        var request = db.transaction(["pokemons"], "readwrite")
                    .objectStore("pokemons")
                    .add(obj);

        // transaction.objectStore(objStoreName).put(obj, key);

        request.onsuccess = function(event) {
            resolve("success");
        };

        request.onerror = function(event) {
            reject("error");
        };
    })

};

export var releasePokemon = function(pokemonId){
    return new Promise((resolve, reject) => {
        var request = db.transaction(["pokemons"], "readwrite")
                    .objectStore("pokemons")
                    .delete(pokemonId);

        request.onsuccess = function(event) {
            resolve("success");
        };

        request.onerror = function(event) {
            reject("error");
        };
    })
}