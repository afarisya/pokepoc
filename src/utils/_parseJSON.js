export function _parseJSON(response) {
    return response.text()
        .then( text => {
            return text ? JSON.parse(text) : {}
    })
}