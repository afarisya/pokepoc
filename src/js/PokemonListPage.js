import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router'; // react-router v4
import { 
    Container,
    Row,
    Col,
    Spinner
} from 'reactstrap';
import queryString from 'query-string'

import store from '../store';

// Components
import PokemonCard from './PokemonCard';
import PokemonPagination from './PokemonPagination';

// Actions
import { reqPokemonList } from '../reducers/PokemonListReducers';
import { reqMyPokemonTotal } from '../reducers/MyPokemonListReducers';

class PokemonListPage extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
            searchQuery: "",
        }
	}

    componentWillMount() {
        store.dispatch(reqMyPokemonTotal());

        this.setState({
            searchQuery: this.props.history.location.search
        }, () => {
            this.renderRoute();
        });
    }

    componentWillReceiveProps(nextProps) {
        // console.log(nextProps)
        if ( this.state.searchQuery !== nextProps.history.location.search ) {
            this.setState({
                searchQuery: nextProps.history.location.search
            }, () => {
                this.renderRoute();
            });
		}
	}

    renderRoute = () => {
        var pageNumber = 1;

        const searchObj = queryString.parse(this.state.searchQuery);

        if ( typeof searchObj.page !== 'undefined' ) {
            pageNumber = parseInt(searchObj.page);
        }

        store.dispatch(reqPokemonList(pageNumber));
    }


	handlePageChange = (pageNumber) => { 
        const searchObj = queryString.parse(this.props.history.location.search);
        searchObj["page"] = pageNumber;
        var searchString = queryString.stringify(searchObj);
        this.props.history.push(this.props.history.location.pathname + "?" + searchString);
    }
    
    render() {
        return (
            <div>
                <Container>
                    { this.props.pokemons.length === 0 ?
                        <Row className="pokemon-list-main-row">
                            <Spinner color="danger" style={{margin: "50% auto"}} />
                        </Row>
                        :
                        <React.Fragment>
                            <div style={{padding: "30px 15px 5px"}}>
                                <h5>All Pokemons (Total {this.props.totalPokemons})</h5>
                                <div>You've catched {this.props.totalCatchedPokemons} of them</div>
                            </div>
                            <Row className="pokemon-list-main-row">
                                {
                                    this.props.pokemons.map((pokemon, index) => {
                                        return (
                                            <Col key={index} className="pokemon-col" xs="6" sm="4" md="3" lg="2">
                                                <PokemonCard id={pokemon.name} name={pokemon.name} />
                                            </Col>
                                        )
                                    })
                                }
                            </Row>
                        </React.Fragment>
                    }
                    <Row style={{padding: "0 15px"}}>
                        <PokemonPagination 
                            activePage={this.props.activePage}
                            totalItems={this.props.totalPokemons}
                            itemsPerPage={this.props.limit}
                            onChange={this.handlePageChange}
                        />
                    </Row>
                </Container>
            </div>
        );
    }
};

function mapStateToProps(state) {
	return {
		activePage      : state.PokemonListReducers.activePage,
		offset          : state.PokemonListReducers.offset,
		limit           : state.PokemonListReducers.limit,
		totalPokemons   : state.PokemonListReducers.totalPokemons,
		pokemons        : state.PokemonListReducers.pokemons,
		totalCatchedPokemons   : state.MyPokemonListReducers.totalPokemons,
	}
}

export default withRouter(connect(mapStateToProps)(PokemonListPage));