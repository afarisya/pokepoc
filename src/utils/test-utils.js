import { ConnectedRouter } from 'react-router-redux';
import { mount } from 'enzyme';
import { mergeDeepRight } from 'ramda';
import React from 'react';
import { Provider } from 'react-redux';
import { createStoreWithMiddleWare, history } from '../store';
import rootReducer from '../reducers';

export const mockData = {
  pokemon: {
    id: 1,
    species: { name: 'bulbasaur' },
    sprites: {
      front_default:
        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/26.png'
    }
  }
};

export const makeMountRender = (Component, defaultProps = {}) => {
  return (customProps = {}) => {
    const props = {
      ...defaultProps,
      ...customProps
    };
    return mount(<Component {...props} />);
  };
};

export const makeStore = (customState = {}) => {
  const root = rootReducer({}, { type: '@@INIT' });
  const state = mergeDeepRight(root, customState);

  return createStoreWithMiddleWare(rootReducer, state);
};

export const reduxify = (Component, props = {}, state = {}) => {
  return function reduxWrap() {
    return (
      <Provider store={makeStore(state)}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </Provider>
    );
  };
};

export const snapshotify = reactWrapper => {
  return reactWrapper.html();
};

export const mocker = apiMock => ({
  fetchRandomPokemon() {
    apiMock
      .onGet(/https:\/\/pokeapi.co\/api\/v2\/pokemon\/\d+/)
      .reply(config => {
        const numberCheck = RegExp(/pokemon\/(\d+)/);
        const pokemonInt = Number(numberCheck.exec(config.url)[1]);

        if (pokemonInt > 151) {
          return [500];
        }

        return [200, mockData.pokemon];
      });

    return this;
  }
});

export const ticks = (callbacks = []) => {
  if (callbacks.length < 1) {
    return null;
  }

  setTimeout(() => {
    const cb = callbacks.shift();

    if (cb) {
      cb();
    }

    ticks(callbacks);
  });
};
