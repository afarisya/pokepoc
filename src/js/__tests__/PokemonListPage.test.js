// PokemonListPage.test.js

import React from 'react';
import { shallow } from 'enzyme';
import { get } from '../../utils/IndexedDb';
import { _parseJSON } from '../../utils/_parseJSON';
import { history } from '../../store'
// require("fake-indexeddb/auto");
import PokemonListPage from '../PokemonListPage';

const fetch = require('node-fetch');

describe("PokemonListPage", () => {
    const page = shallow(
        <PokemonListPage.WrappedComponent 
            history={history} 
            activePage={1}
            offset={0}
            limit={12}
            totalPokemons={0}
            pokemons={[]}
            totalCatchedPokemons={0}
        />
    );

    it("should render pokemon list page if pokemons.length = 0", () => {   
        expect(page.getElements()).toMatchSnapshot();
    });

    it("should show Spinner on pokemon list page if pokemons.length = 0", () => { 
        expect(page.find('Spinner').length).toEqual(1);
    });

    it("should show PokemonPagination on pokemon list page if pokemons.length = 0", () => {  
        expect(page.find('PokemonPagination').length).toEqual(1);
    });

    const page2 = shallow(
        <PokemonListPage.WrappedComponent 
            history={history} 
            activePage={1}
            offset={0}
            limit={12}
            totalPokemons={500}
            pokemons={[{id: "bulbasaur", name: "bulbasaur"}]}
            totalCatchedPokemons={0}
        />
    );

    it("should render pokemon list page if pokemons.length = 1", () => {   
        expect(page2.getElements()).toMatchSnapshot();
    });

    it("should show 1 PokemonCard on pokemon list page if pokemons.length = 1", () => { 
        expect(page2.find('PokemonCard').length).toEqual(1);
    });

    it("should show PokemonPagination on pokemon list page if pokemons.length = 1", () => {  
        expect(page2.find('PokemonPagination').length).toEqual(1);
    });

    // expect(page.prop('activePage')).toEqual(1);

    var targetUrl   = `https://pokeapi.co/api/v2/pokemon?offset=0&limit=12`;

    // Begin fetch 
    fetch(targetUrl)
        .then((response) => {
            if ( response.ok ) {
                return response;
            } else {
                throw Error(response.status);
            }
        })
        .then(response => {
            return _parseJSON(response);
        })
        .then(resp => {
            it("should render pokemon list page with 12 pokemon cards", () => {   
                expect(resp.count).toBe(0);
            });
            // const page3 = shallow(
            //     <PokemonListPage.WrappedComponent 
            //         history={history} 
            //         activePage={1}
            //         offset={0}
            //         limit={12}
            //         totalPokemons={resp.count}
            //         pokemons={resp.results}
            //         totalCatchedPokemons={0}
            //     />
            // );

            // it("should render pokemon list page with 12 pokemon cards", () => {   
            //     expect(page3.getElements()).toMatchSnapshot();
            // });
            // return dispatch(rcvPokemonList(resp));
        })
        // .catch(error => {
        //     console.log(error)
        // });
});

describe('test fetch', () => {
    var count = 0;

    var targetUrl   = `https://pokeapi.co/api/v2/pokemon?offset=0&limit=12`;

    fetch(targetUrl)
        .then((response) => {
            response.json();
        })
        .then(body => {
            count = body.count;
        })
        .then(() => {
            expect(resp.count).toEquals(999);
            
            const page = shallow(
                <PokemonListPage.WrappedComponent 
                    history={history} 
                    activePage={1}
                    offset={0}
                    limit={12}
                    totalPokemons={0}
                    pokemons={[]}
                    totalCatchedPokemons={0}
                />
            );

            it("au amat", () => {   
                expect(page.getElements()).toMatchSnapshot();
            });
        })
        .catch((err) => {

        })
})