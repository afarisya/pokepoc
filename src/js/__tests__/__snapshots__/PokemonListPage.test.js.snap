// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonListPage should render pokemon list page if pokemons.length = 0 1`] = `
Array [
  <div>
    <Container
      tag="div"
    >
      <Row
        className="pokemon-list-main-row"
        tag="div"
      >
        <Spinner
          color="danger"
          style={
            Object {
              "margin": "50% auto",
            }
          }
          tag="div"
          type="border"
        >
          Loading...
        </Spinner>
      </Row>
      <Row
        style={
          Object {
            "padding": "0 15px",
          }
        }
        tag="div"
      >
        <PokemonPagination
          activePage={1}
          itemsPerPage={12}
          onChange={[Function]}
          totalItems={0}
        />
      </Row>
    </Container>
  </div>,
]
`;

exports[`PokemonListPage should render pokemon list page if pokemons.length = 1 1`] = `
Array [
  <div>
    <Container
      tag="div"
    >
      <React.Fragment>
        <div
          style={
            Object {
              "padding": "30px 15px 5px",
            }
          }
        >
          <h5>
            All Pokemons (Total 
            500
            )
          </h5>
          <div>
            You've catched 
            0
             of them
          </div>
        </div>
        <Row
          className="pokemon-list-main-row"
          tag="div"
        >
          <Col
            className="pokemon-col"
            lg="2"
            md="3"
            sm="4"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="6"
          >
            <PokemonCard
              id="bulbasaur"
              name="bulbasaur"
            />
          </Col>
        </Row>
      </React.Fragment>
      <Row
        style={
          Object {
            "padding": "0 15px",
          }
        }
        tag="div"
      >
        <PokemonPagination
          activePage={1}
          itemsPerPage={12}
          onChange={[Function]}
          totalItems={500}
        />
      </Row>
    </Container>
  </div>,
]
`;
